File: backend/src/main/java/com/stockanalysis/backtest/spec/ExitSpec.java
package com.stockanalysis.backtest.spec;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

import java.time.LocalTime;
import java.util.ArrayList;
import java.util.List;

/**
 * Exit rules. Take-profit / stop-loss are percentages from entry price; {@code maxHoldBars}
 * forces a time exit after N bars; {@code closeAtDayEnd} force-closes at the last bar of the
 * trading day (intraday scalping default). Signal conditions may also trigger an exit.
 *
 * <p>{@code bands} optionally overrides take-profit / stop-loss per time of day. The band is
 * resolved against the <em>current</em> bar on every held bar, not against the entry bar, so a
 * position opened at 09:20 switches to the 10:00 band's levels once the clock passes 10:00.
 * First matching band wins; a band that doesn't set a percentage falls back to the base value.
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public class ExitSpec {

    private Double takeProfitPct;
    private Double stopLossPct;
    private Integer maxHoldBars;
    private boolean closeAtDayEnd = true;
    private Logic logic = Logic.OR;
    private List<Condition> conditions = new ArrayList<>();
    private List<TimeBand> bands = new ArrayList<>();

    public ConditionGroup asGroup() {
        return new ConditionGroup(logic, conditions);
    }

    /** Take-profit % in effect at {@code t}: first matching band's value, else the base value. */
    public Double takeProfitPctAt(LocalTime t) {
        TimeBand b = bandAt(t);
        return (b != null && b.getTakeProfitPct() != null) ? b.getTakeProfitPct() : takeProfitPct;
    }

    /** Stop-loss % in effect at {@code t}: first matching band's value, else the base value. */
    public Double stopLossPctAt(LocalTime t) {
        TimeBand b = bandAt(t);
        return (b != null && b.getStopLossPct() != null) ? b.getStopLossPct() : stopLossPct;
    }

    /** Bands are matched in list order, so an earlier entry wins an overlap. */
    public TimeBand bandAt(LocalTime t) {
        if (bands == null || bands.isEmpty() || t == null) {
            return null;
        }
        for (TimeBand b : bands) {
            if (b != null && b.covers(t)) {
                return b;
            }
        }
        return null;
    }

    public Double getTakeProfitPct() {
        return takeProfitPct;
    }

    public void setTakeProfitPct(Double takeProfitPct) {
        this.takeProfitPct = takeProfitPct;
    }

    public Double getStopLossPct() {
        return stopLossPct;
    }

    public void setStopLossPct(Double stopLossPct) {
        this.stopLossPct = stopLossPct;
    }

    public Integer getMaxHoldBars() {
        return maxHoldBars;
    }

    public void setMaxHoldBars(Integer maxHoldBars) {
        this.maxHoldBars = maxHoldBars;
    }

    public boolean isCloseAtDayEnd() {
        return closeAtDayEnd;
    }

    public void setCloseAtDayEnd(boolean closeAtDayEnd) {
        this.closeAtDayEnd = closeAtDayEnd;
    }

    public Logic getLogic() {
        return logic;
    }

    public void setLogic(Logic logic) {
        this.logic = logic == null ? Logic.OR : logic;
    }

    public List<Condition> getConditions() {
        return conditions;
    }

    public void setConditions(List<Condition> conditions) {
        this.conditions = conditions;
    }

    public List<TimeBand> getBands() {
        return bands;
    }

    public void setBands(List<TimeBand> bands) {
        this.bands = bands == null ? new ArrayList<>() : bands;
    }
}

File: backend/src/main/kotlin/com/stockanalysis/backtest/spec/ExitSpec.kt
package com.stockanalysis.backtest.spec

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonSetter
import com.fasterxml.jackson.annotation.Nulls
import java.time.LocalTime

/**
 * 매도(청산) 규칙.
 *
 * 익절·손절은 **진입가 대비 퍼센트**이고, [maxHoldBars]는 N봉이 지나면 시간 청산,
 * [closeAtDayEnd]는 장 마지막 봉에서 강제 청산입니다(단타 기본값). 지표 조건([conditions])도
 * 매도 트리거가 됩니다.
 *
 * [bands]는 시간대별로 익절·손절을 덮어씁니다. **보유 중인 매 봉마다 그 봉의 시각으로 다시**
 * 판정하므로 시계가 다음 밴드로 넘어가면 선도 따라 움직입니다. 먼저 맞는 밴드가 이기니까
 * 겹쳐도 되고, 좁은 구간을 앞에 두면 됩니다.
 */
@JsonIgnoreProperties(ignoreUnknown = true)
class ExitSpec {

    var takeProfitPct: Double? = null
    var stopLossPct: Double? = null
    var maxHoldBars: Int? = null

    @JsonSetter(nulls = Nulls.SKIP)
    var closeAtDayEnd: Boolean = true

    @JsonSetter(nulls = Nulls.SKIP)
    var logic: Logic = Logic.OR

    @JsonSetter(nulls = Nulls.AS_EMPTY)
    var conditions: MutableList<Condition> = mutableListOf()

    /** 시간대 밴드는 나중에 추가된 필드라 예전 행에는 없습니다 — null이면 빈 목록으로 읽습니다. */
    @JsonSetter(nulls = Nulls.AS_EMPTY)
    var bands: MutableList<TimeBand> = mutableListOf()

    /** 매도 조건들을 그룹으로 감싼 것. 엔진이 매수 조건과 똑같은 방식으로 평가합니다. */
    fun asGroup(): ConditionGroup = ConditionGroup(logic, conditions)

    /** `t` 시점에 유효한 익절 % — 먼저 맞는 밴드의 값, 없으면 기본값. */
    fun takeProfitPctAt(t: LocalTime?): Double? {
        val band = bandAt(t)
        return band?.takeProfitPct ?: takeProfitPct
    }

    /** `t` 시점에 유효한 손절 % — 먼저 맞는 밴드의 값, 없으면 기본값. */
    fun stopLossPctAt(t: LocalTime?): Double? {
        val band = bandAt(t)
        return band?.stopLossPct ?: stopLossPct
    }

    /** 목록 순서대로 찾으므로 구간이 겹치면 앞에 있는 것이 이깁니다. */
    fun bandAt(t: LocalTime?): TimeBand? {
        if (bands.isEmpty() || t == null) {
            return null
        }
        return bands.firstOrNull { it.covers(t) }
    }
}
